exporter/splunkhecexporter/client.go

```diff
@@ -153,21 +157,116 @@ func (c *client) sendSplunkEvents(ctx context.Context, splunkEvents []*splunk.Ev
 	return nil
 }
 
+func subLog(ld pdata.Logs, logIdx logIndex) pdata.Logs {
+	if logIdx.zero() {
+		return ld
+	}
+	clone := ld.Clone().InternalRep()
+
+	subset := *clone.Orig
+	subset = subset[logIdx.origIdx:]
+	subset[0].InstrumentationLibraryLogs = subset[0].InstrumentationLibraryLogs[logIdx.instIdx:]
+	subset[0].InstrumentationLibraryLogs[0].Logs = subset[0].InstrumentationLibraryLogs[0].Logs[logIdx.logsIdx:]
+
+	clone.Orig = &subset
+	return pdata.LogsFromInternalRep(clone)
+}
+
+func (c *client) chunk(ctx context.Context, ld pdata.Logs, max int, send func(ctx context.Context, buf *bytes.Buffer) error) (numDroppedLogs int, err error) {
+	buf := bytes.NewBuffer(make([]byte, 0, max))
+	encoder := json.NewEncoder(buf)
+	checkPoint := 0
+	var logIdx logIndex
+
+	rls := ld.ResourceLogs()
+	for i := 0; i < rls.Len(); i++ {
+		ills := rls.At(i).InstrumentationLibraryLogs()
+		for j := 0; j < ills.Len(); j++ {
+			logs := ills.At(j).Logs()
+			for k := 0; k < logs.Len(); k++ {
+				event := mapLogRecordToSplunkEvent(logs.At(k), c.config, c.logger)
+				if err := encoder.Encode(event); err != nil {
+					numDroppedLogs++
+					continue
+				}
+
+				buf.WriteString("\r\n\r\n")
+
+				if buf.Len() < max {
+					checkPoint = buf.Len()
+					continue
+				}
+
+				sending := buf.Next(checkPoint)
+				leaving := buf.Bytes()
+
+				if err := send(ctx, bytes.NewBuffer(sending)); err != nil {
+					return numDroppedLogs, consumererror.PartialLogsError(err, subLog(ld, logIdx))
+				}
+
+				logIdx = logIndex{origIdx: i, instIdx: j, logsIdx: k,}
+
+				buf.Reset()
+				buf.Write(leaving)
+
+				checkPoint = 0
+			}
+		}
+	}
+
+	if buf.Len() > 0 {
+		if err := send(ctx, buf); err != nil {
+			return numDroppedLogs, consumererror.PartialLogsError(err, subLog(ld, logIdx))
+		}
+	}
+
+	return numDroppedLogs, nil
+}
+
 func (c *client) pushLogData(ctx context.Context, ld pdata.Logs) (numDroppedLogs int, err error) {
 	c.wg.Add(1)
 	defer c.wg.Done()
 
-	splunkEvents := logDataToSplunk(c.logger, ld, c.config)
-	if len(splunkEvents) == 0 {
-		return 0, nil
-	}
+	gzipWriter := c.zippers.Get().(*gzip.Writer)
+	defer c.zippers.Put(gzipWriter)
 
-	err = c.sendSplunkEvents(ctx, splunkEvents)
-	if err != nil {
-		return ld.LogRecordCount(), err
+	gzipBuf := bytes.NewBuffer(make([]byte, 0, c.config.MaxContentLengthLogs))
+	gzipWriter.Reset(gzipBuf)
+	defer gzipWriter.Close()
+
+	// Callback when each chunk is to be sent.
+	sendBatch := func(ctx context.Context, buf *bytes.Buffer) error {
+		compression := false
+		var reader io.Reader
+
+		if buf.Len() < 1500 || c.config.DisableCompression {
+			compression = false
+			reader = buf
+		} else {
+			compression = true
+
+			gzipBuf.Reset()
+			gzipWriter.Reset(gzipBuf)
+
+			if _, err := io.Copy(gzipWriter, buf); err != nil {
+				return err
```

Comment:
I'd recommend adding more context to all of the `return err` throughout, ie `fmt.Errorf("failed copying buffer to gzip writer: %v", err)` because it'll be difficult to track down the exact line returning the error.